server/src/main/java/org/elasticsearch/action/admin/indices/resolve/TransportResolveClusterAction.java

```diff
@@ -0,0 +1,286 @@
+                            return;
+                        }
+                        ResolveClusterInfo info = response.getResolveClusterInfo().get(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY);
+                        if (info != null) {
+                            clusterInfoMap.put(
+                                clusterAlias,
+                                new ResolveClusterInfo(info.isConnected(), skipUnavailable, info.getMatchingIndices(), info.getBuild())
+                            );
+                        }
+                        if (resolveClusterTask.isCancelled()) {
+                            releaseResourcesOnCancel.run();
+                        }
+                    }
+
+                    @Override
+                    public void onFailure(Exception failure) {
+                        if (resolveClusterTask.isCancelled()) {
+                            releaseResourcesOnCancel.run();
+                            return;
+                        }
+                        System.err.println(">>> HHHHH failure FAILURE: " + failure.getClass() + "; " + failure);
+                        if (notConnectedError(failure)) {
+                            clusterInfoMap.put(clusterAlias, new ResolveClusterInfo(false, skipUnavailable));
+                        } else if (ExceptionsHelper.unwrap(
+                            failure,
+                            ElasticsearchSecurityException.class
+                        ) instanceof ElasticsearchSecurityException ese) {
+                            clusterInfoMap.put(clusterAlias, new ResolveClusterInfo(true, skipUnavailable, ese.getMessage()));
+                        } else if (ExceptionsHelper.unwrap(failure, IndexNotFoundException.class) instanceof IndexNotFoundException infe) {
+                            System.err.println(">>>> HHH INDEX NOT FOUND!!! <<<<<<<<<<<<<<<");
+                            clusterInfoMap.put(clusterAlias, new ResolveClusterInfo(true, skipUnavailable, infe.getMessage()));
+                        } else {
+                            Throwable cause = ExceptionsHelper.unwrapCause(failure);
+                            // when querying an older cluster that does not have the _resolve/cluster endpoint,
+                            // we will get this error at the Transport layer, since there are version guards
+                            // on the Writeables for this Action.
+                            if (cause instanceof UnsupportedOperationException
+                                && cause.getMessage().contains("ResolveClusterAction requires at least Transport Version")) {
+                                // Since this cluster does not have _resolve/cluster, we will call the _resolve/index
+                                // endpoint to fill in the matching_indices field of the response for that cluster
+                                ResolveIndexAction.Request resolveIndexRequest = new ResolveIndexAction.Request(
```

Comment:
That;s a pretty nice fallaback: is the result from it as accurate as the original behaviour?

Reply:
For the `matching_indices` field yes, since under the hood `_resolve/cluster` uses the ResolveIndexAction to determine matching_indices, so it is the same code doing the check.

What's absent though is that I don't get back the cluster `version` info from the remote cluster when calling `_resolve/index`. I tried to do a second fallback for that by calling the `GET /` endpoint of the remote cluster, but that "Main" code is not in `server` code module, it is in the `./modules/rest-root` and this code is unable to access it (using the typical `remoteClusterClient.admin().indices().execute()`). 

Could we move the `MainRest` code to `server`?

Reply:
The `main` action is kinda deprecated and won't be available in serverless, we shouldn't be relying on it here.

Reply:
OK, but providing cluster version info is pretty useful for CCS, even for serverless, because even though serverless is always "latest" version, the `version.build_flavor` field reports `serverless` from `GET /`, so that is useful info in a CCS scenario (once we allow CCS between on-prem, ESS and serverless). Eventually we may find the Feature API to be even more useful, but that future work, possibly well into the future.

Reply:
Can we avoid exposing version information at all? Why does Kibana need the Elasticsearch version? If it's for feature checks, we need to start rethinking that type of behavior since it won't work in serverless.

Reply:
Version information about remote clusters has been asked for by CCS customers and is on the CCS roadmap (https://elasticco.atlassian.net/jira/polaris/projects/EPD/ideas/view/3054449?selectedIssue=EPD-28 and https://elasticco.atlassian.net/jira/polaris/projects/EPD/ideas/view/3054449?selectedIssue=EPD-51).

CCS will continue to be heavily used by non-serverless customers. There is also a need to do CCS between on-prem and serverless.

As I said above the Build version info for serverless identifies it as "serverless" in the `version.build_flavor`, so that is the important part for serverless, not the actual version (unlike on-prem and ESS).

So it seems to me that version will continue to be important for the foreseeable future for CCS customers.

Long term we likely want to move to the Feature API, but even after we figure out how to support that, we know that many customers lag behind in upgrading to latest versions and the CCS roadmap also includes want to support CCS across a wider array of older versions: https://elasticco.atlassian.net/jira/polaris/projects/EPD/ideas/view/3054449?selectedIssue=EPD-51

All that said, it isn't essential to include version right now in this endpoint for the initial case that Kibana has, but it would be useful additional info to show to users/admins when determining how to construct cross-cluster searches (or diagnose why a particular attempt failed).

@tylerperk - What are your thoughts?

Reply:
@quux00 I think you summarised it well, thanks. This version info does indeed help us satisfy multiple user and stakeholder requests in our roadmap. If there isn't a strong objection to including version info now, I'd prefer to keep it, as then we start the clock on versions that can provide this info, whereas if we decide to add it later then only even newer versions can provide it.

Reply:
> There is also a need to do CCS between on-prem and serverless.

> As I said above the Build version info for serverless identifies it as "serverless" in the version.build_flavor, so that is the important part for serverless, not the actual version (unlike on-prem and ESS

How do you envision this working when serverless exposes no version information?

I'm hesitant to add such a new API exposing version information because it is going in the wrong direction. I understand there are immediate needs for CCS now, but I am concerned some may think this new API "just" needs to be exposed in serverless to make CCS work there, yet it cannot be exposed with version information (which appears crucial to its design).

Reply:
I think for serverless, all we need to expose is that the version is "serverless". 

> How do you envision this working when serverless exposes no version information?

Can you elaborate more what that means? When I tested against a serverless project a couple of weeks ago, the `GET /` endpoint was available and the Build-version info told it was serverless by means of `version.build_flavor=serverless`. Since there is only one serverless version (roughly "latest"), that is all a CCS user would need to know to query it, I believe. The concern is with older versions.

Do you mean that the `GET /` API will be removed from serverless? If yes, will there be some other endpoint to tell us that this is a serverless deployment? I would think we need to provide some API to indicate that information.

Longer term we would like to leverage the Features API, but how exactly that will be done and whether Features will be granular enough is still TBD.

Reply:
CCS currently has arbitrary compatibility bounds on the previous minor version of ES. If you know a cluster is "serverless", how do you know it fits within those bounds?

> Do you mean that the GET / API will be removed from serverless?

It will likely not be removed, but it should not be relied upon. You can get version information from the nodes info api.

Reply:
I got some new (to me) information from @tylerperk that may help decide on how to proceed here - namely, that senior/executive management in Elastic have recently decided that we will not support (existing) cross-cluster search to serverless. Serverless will look into some future design model (I don't have any details on that). In addition, this new `_resolve/cluster` endpoint is also not going to be available in serverless.

So in that context, let me rephrase the proposal/argument for having version in this endpoint and see if we can come to consensus.

The use case for this endpoint is for Kibana (and other any cross-cluster search user) to have a single endpoint to hit that tells them as much information as possible about whether a cross-cluster search they are going to do will work.

To that end, this endpoint accepts the same index expression you would provide to a cross-cluster search and gives you:

* up-to-date information of whether each remote cluster included is connected
* the skip_unavailable status of each cluster (important for determining if a whole search will fail if the search on that cluster fails)
* whether the index expression for each cluster is likely to match anything (at an index level, not a document/aggs level) - this includes considering whether the index is closed or not
* detect whether an error will occur on each given clusters - errors that can be detected by this endpoint are:
  * index not found
  * security/permission issues
* the ES version that each cluster is running

In the future, we can expand this to include additional useful information, such as a list of Features that each remote includes, once we have defined if and how cross-cluster search will utilize the Features API.

Kibana intends to display all this new information, including ES version of each cluster, on a new UI soon to be developed. The ways in which the ES version information is useful are:

1) Elasticsearch admins have no easy way to get information of what version remote clusters are running since in many cases they are run by different sub-organizations in the company and there is no current endpoint in Elasticsearch that supports cross-cluster version info display. The `GET /` endpoint and the nodes info endpoint do not work cross-cluster.

2) Since officially Elasticsearch only supports cross-cluster search back to the previous minor, this information is useful to users to decide whether to deselect older clusters from a cross-cluster search. Kibana will be building a UI to support selecting (or unselecting) which clusters to include in a search.

So that is why we have implemented this first version of the endpoint with ES version included.